### PR TITLE
Add ServiceHubContextBuilder guide

### PR DESCRIPTION
**Pull request by Y-Sindo**

### CODE REVIEW

File: docs/service-hub-context-builder-guide.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
+<!-- This doc is intent to replace the old management-sdk-guide -->
```
Comment (vicancy) [inline docs/service-hub-context-builder-guide.md:1]: directly change the management sdk guide, mark the SDK version for the old usage, and add a migration guide?

File: docs/management-sdk-migration.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
+# `ServiceManagerBuilder` to `ServiceHubContextBuilder` Migration guidance
```
Comment (wanlwanl) [inline docs/management-sdk-migration.md:1]: guidance -> Guidance

File: docs/management-sdk-migration.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+This guidance is intended to assist in the migration to `ServiceHubContextBuilder` from `ServiceManagerBuilder`. It will focus on side-by-side comparisons for similar operations between the legacy and the new APIs. The new APIs will be released in version 10.0.0 .
+
+We assume that you are familiar with the usage of `ServiceManagerBuilder` and the related APIs, otherwise, please refer the [Azure SignalR Service Management SDK ](management-sdk-guide.md) other than this guidance.
+- [`ServiceManagerBuilder` to `ServiceHubContextBuilder` Migration guidance](#servicemanagerbuilder-to-servicehubcontextbuilder-migration-guidance)
+  - [Migration Benefits](#migration-benefits)
+  - [Change Overview](#change-overview)
+  - [API comparisons](#api-comparisons)
+    - [Entry point](#entry-point)
+    - [Negotiate](#negotiate)
```
Comment (wanlwanl) [inline docs/management-sdk-migration.md:11]: Negotiate -> Negotiation ?

File: docs/management-sdk-migration.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    - [Send Messages and Manage Groups](#send-messages-and-manage-groups)
+    - [Health check](#health-check)
+
+
+
+## Migration Benefits
+* The new APIs provide more functionalities to manage your clients and groups, such as closing a connection by connection id, checking if a connection exists, if a user exists, if a group exists. 
+* The new APIs provide more options for negotiation, such as whether the client is a diagnostic client.
+* The new APIs are more friendly for negotiation with multiple SignalR Service instances. `IServiceManager.GetClientEndpoint` and `IServiceManager.GenerateClientAccessToken` are combined into one method to make sure the client endpoint and the access token come from the same SignalR Service endpoint. An `HttpContext` instance is allowed to passed into the endpoint router to provide more information for the routing. <!--Todo Add link about sharding doc-->
```
Comment (wanlwanl) [inline docs/management-sdk-migration.md:20]: to passed -> to pass

Comment (wanlwanl) [reply docs/management-sdk-migration.md:20]: `<!--Todo Add link about sharding doc-->` also include in this PR?

File: docs/management-sdk-migration.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+## Change Overview
+The following image shows the relation for the legacy APIs (green blocks) and the new APIs (blue blocks). Currently, to get a `ServiceHubContext` which implements `IServiceHubContext` interface, you just need to create it directly from `ServiceHubContextBuilder`. The `ServiceHubContext` abstract class combines the functionalities of `IServiceManager` and `IServiceHubContext`.
+
+![image](./images/migration-class-diagram.png)
```
Comment (wanlwanl) [inline docs/management-sdk-migration.md:25]: can it moves to the middle in width?

File: docs/management-sdk-migration.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+### Send Messages and Manage Groups
+**Legacy APIs**
+```cs
+try
+{
+    var hubcontext = await serviceManager.CreateHubContextAsync("<Your Hub Name>");
+
+    // Broadcast
+    hubContext.Clients.All.SendAsync(callbackName, obj1, obj2, ...);
```
Comment (wanlwanl) [inline docs/management-sdk-migration.md:74]: await ...

File: docs/management-sdk-migration.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    await hubContext.DisposeAsync();
+}
+```
+
+**New APIs**
+```cs
+try
+{
+    // Broadcast
+    hubContext.Clients.All.SendAsync(callbackName, obj1, obj2, ...);
```
Comment (wanlwanl) [inline docs/management-sdk-migration.md:92]: await ...

File: docs/management-sdk-migration.md (hunk 5 above)
Comment (Y-Sindo) [reply docs/management-sdk-migration.md:25]: We cannot do that only with markdown.

File: docs/management-sdk-migration.md (hunk 6 above)
Comment (Y-Sindo) [reply docs/management-sdk-migration.md:74]: Haha I copied it from the previous guide and didn't notice this.

File: docs/management-sdk-migration.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+**Legacy APIs**
+```cs
+var isHealthy = await serviceManager.IsServiceHealthy(cancellationToken);
+```
+
+**New APIs**
+
+N/A
+
+Currently in new APIs, you can't check service health directly. But the health check is integrated into negotiation API partially (implemented in [Persistent mode](management-sdk-guide.md#transport-type) but not yet for [Transient mode](management-sdk-guide.md#transport-type). If you need to check the service health yourself, you could call our [health check REST API](https://docs.microsoft.com/en-us/azure/azure-signalr/signalr-quickstart-rest-api#service-health) yourself.
```
Comment (vicancy) [inline docs/management-sdk-migration.md:115]: can we add the support for transient mode?

File: docs/management-sdk-migration.md (hunk 5 above)
Comment (wanlwanl) [reply docs/management-sdk-migration.md:25]: ok

File: docs/management-sdk-migration.md (hunk 4 above)
Comment (Y-Sindo) [reply docs/management-sdk-migration.md]: Sharding doc is not ready. It has to wait for this doc.

File: docs/management-sdk-migration.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    // add user to group
+    await hubContext.UserGroups.AddToGroupAsync(userId, groupName);
+}
+finally
+{
+    await hubContext.DisposeAsync();
+}
+```
+
+With legacy APIs, you need to create a hub context from `IServiceManager`, but with new APIs, you have built a hub context from `ServiceHubContextBuilder` at the beginning. The usages of hub context are nearly the same, except that in the legacy APIs you get an interface `IServiceHubContext` while in the new APIs you get an abstract class `ServiceHubContext`.
```
Comment (vicancy) [inline docs/management-sdk-migration.md:113]: > With legacy APIs, you need to create a hub context from `IServiceManager`, but with new APIs, you have built a hub context from `ServiceHubContextBuilder` at the beginning.

why the legacy one can't create the hub context from the beginning?

File: docs/management-sdk-migration.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    .WithLoggerFactory(loggerFactory) // optional, specify a custom logger factory instance for the whole management SDK
+    .BuildServiceManager(); // note this
+```
+
+<!--Add sharding link-->
+### Create service hub context
+
+**Legacy APIs**
+```cs
+var serviceHubContext = await serviceManager.CreateServiceHubContextAsync("<Your Hub Name>", loggerFactory, cancellationToken);
```
Comment (vicancy) [inline docs/management-sdk-migration.md:49]: question: L81 has a `CreateHubContextAsync`, here has a `CreateServiceHubContextAsync`, are these 2 different methods creating different objects?

File: docs/management-sdk-migration.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
+# `IServiceManager` to `ServiceManager` Migration Guidance
+
+This guidance is intended to assist in the migration to `ServiceManager` from `IServiceManager`. It will focus on side-by-side comparisons for similar operations between the legacy and the new APIs. The new APIs will be released in version 10.0.0 .
```
[the remark this answers is not in the document]
Comment (Y-Sindo) [reply docs/management-sdk-migration.md:3]: The last `.` is a period.

File: docs/management-sdk-migration.md (hunk 10 above)
Comment (Y-Sindo) [reply docs/management-sdk-migration.md]: They are error. They should be both `CreateHubContextAsync`

File: docs/management-sdk-migration.md (hunk 9 above)
Comment (Y-Sindo) [reply docs/management-sdk-migration.md]: Outdated.

File: docs/management-sdk-migration.md [hunk 12]
```diff
@@ -1,6 +1,6 @@
 # `IServiceManager` to `ServiceManager` Migration Guidance
 
-This guidance is intended to assist in the migration to `ServiceManager` from `IServiceManager`. It will focus on side-by-side comparisons for similar operations between the legacy and the new APIs. The new APIs will be released in version 10.0.0 .
+This guidance is intended to assist in the migration to `ServiceManager` from `IServiceManager`. It will focus on side-by-side comparisons for similar operations between the legacy and the new APIs. The new APIs will be released in 10.0.0 version.
```
Comment (vicancy) [inline docs/management-sdk-migration.md]: we don't have version `10.0.0`

**Review by vicancy — APPROVED** [review]
Looks pretty nice to me.

File: docs/management-sdk-guide.md [hunk 13]
```diff
@@ -122,7 +122,7 @@ var connection = new HubConnectionBuilder().WithUrl("http://<Your Host Name>/<Yo
 await connection.StartAsync();
 ```
 
-Please note that by default we have a mechanism to check if your SignalR Service is healthy. If none of your SignalR Service is healthy during negotiation, then an `AzureSignalRNotConnectedException` is thrown.
+<!-- Please note that by default we have a mechanism to check if your SignalR Service is healthy. If none of your SignalR Service is healthy during negotiation, then an `AzureSignalRNotConnectedException` is thrown. -->
```
Comment (vicancy) [inline docs/management-sdk-guide.md:125]: just delete?

**Review by vicancy — APPROVED** [review]

**Pull request merged**